Add checkmark next to replicated/backed up files

tw4l opened the pull request:
Fixes #1291

Please feel free to suggest different placement/styles or make changes yourself! I tried a few different placements and settled on this for now but I don't have the design eye some of you have :)

We may want to make profiles more consistent with files in exposing a `numReplicas` int rather than the `replicas` array but it would involve some more extensive changes to profiles backend so it seemed out of scope for this PR.

## Screenshots

### Crawl

<img width="1130" alt="Screen Shot 2023-11-07 at 4 20 02 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/fce351af-a42d-493a-9813-d109898d1be3">

### Upload

<img width="1128" alt="Screen Shot 2023-11-07 at 4 20 28 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/3a73aff1-7e9f-4fa8-9fd7-9d119b331acd">


### Crawl/Upload Tooltip

<img width="147" alt="Screen Shot 2023-11-07 at 4 20 09 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/27c987fe-3fe3-43a3-9c9c-9d829e18a842">


### Browser Profile

<img width="1128" alt="Screen Shot 2023-11-07 at 4 20 38 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/5425f1ba-89cf-45b1-9f99-28d9ebff5753">

### Browser Profile Tooltip

<img width="160" alt="Screen Shot 2023-11-07 at 4 20 55 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/9286da66-9981-42e3-bd19-e74335ae1547">

tw4l committed:
Add cloud-check icon next to replicated archived item files

tw4l committed:
Add cloud-check for replicated browser profiles in list

emma-sg commented:
Maybe out of scope for this PR, but should the "Download Files" header just say "Files" if that view is used to look at the files at exist (and their status), not just download them? cc @Shrinks99

emma-sg commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+              ${when(
+                data.resource && data.resource.replicas.length > 0,
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
+                  <sl-icon
```
Alignment of the icon seems off in the screenshot of this — I believe the bottom of the icon should align vertically with the text baseline?

tw4l replied:
Fixed!

tw4l committed:
Code review revisions

- Fix alignment
- Move archived items icon to left of file size
- Rename Download Files tab to Files

tw4l committed:
Add more left padding before icon

tw4l commented:
@emma-sg File tab renamed, alignment fixed, icons for archived items moved left of file size, and screenshots updated! Should be good for another look :)

emma-sg reviewed (changes requested):
Better, but it's still misaligned here: 
<img width="453" alt="Screenshot 2023-11-07 at 4 33 29 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5727389/531ea389-6ae0-4854-a8c4-92ee47659970">

An alignment of `baseline` (I think class `items-baseline`) might work, rather than `end`?

Edit: tried `baseline` and it made it worse — lemme poke around a little and see what I can figure out.

I'm also noticing the tooltips are misaligned horizontally, but I'm not sure how much control you have over that without digging deeper that we really need to here.

emma-sg commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -797,8 +794,19 @@ ${this.crawl?.description}
+                        >
+                          <sl-icon
+                            name="cloud-check"
+                            class="h-4 pr-2 shrink-0 text-success"
```
I think this should fix the tooltip alignment issue — `pr` is padding-right, which is inside the box, whereas `mr` is margin-right, which is outside the box. The tooltip is seeing two extra pixels of width inside the box it's positioning itself relative to, and instead we just want to push the whole box over 2 units (quarters of a `rem` i think?) without making it wider.
```suggestion
                            class="h-4 mr-2 shrink-0 text-success"
```

emma-sg commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
+                  <sl-icon
+                    name="cloud-check"
+                    class="pl-2 text-success"
```
```suggestion
                    class="ml-2 text-success"
```

emma-sg reviewed (changes requested):
Figured out the tooltip alignment!

emma-sg commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
         <div class="grid grid-cols-8 gap-3 md:gap-5" role="row">
           <div class="col-span-8 md:col-span-3 p-2" role="cell">
-            <div class="font-medium mb-1">${data.name}</div>
+            <div class="font-medium flex items-end">
```
```suggestion
            <div class="font-medium text-sm">
```
The `text-sm` here aligns the text vertically in the container correctly:

With:
![localhost_9870_orgs_default-org_browser-profiles](https://github.com/webrecorder/browsertrix-cloud/assets/5727389/066f1486-3d08-4599-8344-bd8e222f8608)

Without:
![localhost_9870_orgs_default-org_browser-profiles (1)](https://github.com/webrecorder/browsertrix-cloud/assets/5727389/43bdb60b-9c04-4ca9-9aed-3579cb5ac6b3)

emma-sg commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -797,8 +794,19 @@ ${this.crawl?.description}
                     </div>
                     <div
-                      class="whitespace-nowrap text-sm font-mono text-neutral-400"
+                      class="whitespace-nowrap text-sm font-mono text-neutral-400 flex items-end"
```
```suggestion
                      class="whitespace-nowrap text-sm font-mono text-neutral-400"
```

emma-sg commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -797,8 +794,19 @@ ${this.crawl?.description}
+                        >
+                          <sl-icon
+                            name="cloud-check"
+                            class="h-4 pr-2 shrink-0 text-success"
```
```suggestion
                            class="w-4 h-4 align-text-bottom pr-2 shrink-0 text-success"
```

tw4l replied:
Multiple suggestions per line led to this suggestion being closed, so implementing separately

emma-sg commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
+                  <sl-icon
+                    name="cloud-check"
+                    class="pl-2 text-success"
```
```suggestion
                    class="w-4 h-4 align-text-bottom pl-2 text-success"
```

tw4l replied:
Multiple suggestions per line led to this suggestion being closed, so implementing separately

emma-sg reviewed (changes requested):
And here's vertical alignment! This makes the icons 16x16px, which is their native size, and removes the flex positioning.

tw4l committed:
Change padding to margin

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

tw4l committed:
Change padding to margin

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

tw4l committed:
Change padding-right to margin-right

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

tw4l committed:
Remove flex alignment

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

tw4l commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -799,6 +796,17 @@ ${this.crawl?.description}
+                        >
+                          <sl-icon
+                            name="cloud-check"
+                            class="h-4 mr-2 shrink-0 text-success"
```
```suggestion
                            class="w-4 h-4 mr-2 align-text-bottom shrink-0 text-success"
```

tw4l commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
+                  <sl-icon
+                    name="cloud-check"
+                    class="ml-2 text-success"
```
```suggestion
                    class="w-4 h-4 ml-2 align-text-bottom text-success"
```

tw4l committed:
Apply suggestions from code review

ikreymer commented:
Some minor UI thoughts, otherwise all looks good
- What about using the two cloud `clouds` icon from: https://shoelace.style/components/icon to indicate that its replicated. (Since its always already in one cloud, heh)
- What about putting icon on the left somewhere - maybe next to the file icon (and adding padding otherwise if not replicated). Just seems easier to see, since size is all the way on other side, though maybe we solve that by making the row smaller.

Don't feel too strongly about any of this, just some thoughts :)
Otherwise looks good!

Shrinks99 commented:
No super strong feelings on the cloud icon, would be fine to switch to `clouds`, could be neat!

EDIT: Tried it locally, I dig the clouds!  Pretty sure we'll never use these in the context of weather so good to use here :)

For other areas of the app (org dashboard, archived item type / status) when using icons alongside text the leftmost icon is used to indicate the type of item.  Here that is (admititdly somewhat redundantly) always a WACZ using the ZIP file icon, but generally I like the placement of the replication icon next to the file size.  Feels consistent for what we're trying to convey!

---

"File Replicated" is not descriptive enough for me.  "Replicated to what?  Replicated how many times? (might address in the future)  What is the significance of files being replicated?" are all questions I'd expect from users with this language.  I think "File Backed Up" is probably more accessible with "n File Backups" being also good if we decide to add a count later if there are more than 1 replicas set up?  IMO the concept of "replication" is a little more open ended than the concept of "backups" which is what we're trying to convey with this feature?  For a user-driven replication from our service to their S3 bucket we should use different language, "replicated" may be a better fit for that.  Open to any alternate less technical language suggestions!

Approved pending suggestions / further consideration of language.

Shrinks99 commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+                data.resource && data.resource.replicas.length > 0,
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
+                  <sl-icon
+                    name="cloud-check"
```
Heres a suggestion for changing to [`clouds`](https://icons.getbootstrap.com/icons/clouds/) :)

```suggestion
                    name="clouds"
```

Shrinks99 commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -799,6 +796,17 @@ ${this.crawl?.description}
+                      ${when(
+                        file.numReplicas > 0,
+                        () => html` <sl-tooltip
+                          content=${msg("File replicated")}
```
```suggestion
                          content=${msg("File backed up")}
```

Shrinks99 commented on frontend/src/pages/org/browser-profiles-list.ts:
```diff
@@ -101,7 +102,18 @@ export class BrowserProfilesList extends LiteElement {
+              <span>${data.name}</span>
+              ${when(
+                data.resource && data.resource.replicas.length > 0,
+                () => html` <sl-tooltip content=${msg("Profile replicated")}>
```
```suggestion
                () => html` <sl-tooltip content=${msg("Profile backed up")}>
```

tw4l commented:
> "File Replicated" is not descriptive enough for me. "Replicated to what? Replicated how many times? (might address in the future) What is the significance of files being replicated?" are all questions I'd expect from users with this language. I think "File Backed Up" is probably more accessible with "n File Backups" being also good if we decide to add a count later if there are more than 1 replicas set up?

We can already say `n file backups`/`n replicas` if you prefer that language! We'll only have the one default replica (aka backup) enabled at first, but when we allow people to bring their own S3 buckets that will likely change.

> IMO the concept of "replication" is a little more open ended than the concept of "backups" which is what we're trying to convey with this feature? For a user-driven replication from our service to their S3 bucket we should use different language, "replicated" may be a better fit for that. Open to any alternate less technical language suggestions!

The mechanism we're using for the Browsertrix Cloud service's backup and for user-driven replication to additional S3 buckets is/will be identical. Our motivation for getting this feature in quickly was to have backups for the hosted service, but the way that works is the same as replication to additional s3 buckets will be. So I'm not sure we can use different language for the two, hence why I went with "replicated" instead of "backed up".

I can live with the back up language though! That's essentially what replicas are.

I initially suggested `Backed up to n locations` or `n file backups`, but I guess if the service will only have one replica to begin with, maybe just `Backed up` is fine for now?

tw4l committed:
Use clouds icon

tw4l committed:
Change tooltips to 'Backed up' for now

Shrinks99 commented:
> The mechanism we're using for the Browsertrix Cloud service's backup and for user-driven replication to additional S3 buckets is/will be identical.

This makes lots of sense from a back-end perspective as they're the same process. I think we could handle what I'm talking about on the front end with a few checks depending on what / how we allow users to replicate data or choose their back end storage solution?  The semantics of "Webrecorder has made a second copy of your data for you",  "Your data has been backed up to your storage servers", and "your item has been transferred entirely to your storage servers" are quite different and I'd want to convey that with different icons and text to make it clear to users who has what and where!

Can we leave as "backed up" for now and return to this when it's time to do that?

tw4l commented:
> > The mechanism we're using for the Browsertrix Cloud service's backup and for user-driven replication to additional S3 buckets is/will be identical.
> 
> This makes lots of sense from a back-end perspective as they're the same process. I think we could handle what I'm talking about on the front end with a few checks depending on what / how we allow users to replicate data or choose their back end storage solution? The semantics of "Webrecorder has made a second copy of your data for you", "Your data has been backed up to your storage servers", and "your item has been transferred entirely to your storage servers" are quite different and I'd want to convey that with different icons and text to make it clear to users who has what and where!
> 
> Can we leave as "backed up" for now and return to this when it's time to do that?

All very reasonable and sounds good to me! Merging this now then :)
